### PR TITLE
`<rel>` and `<unit>` theme inheritance

### LINKED ISSUE #3951: Suggestion: `rel`-class can inherit value from `unit`-class theme elements
I know `rel` objects are not really supported as theme elements on their own, but since the `size` argument in `element_line/rect/text()` can take a `rel` object and inherit the (modified) value from the parent, it doesn't seem like too far a stretch to have child `rel` theme elements.

Suppose you'd like your legend keys to be half as tall as usual, it would be neat if you could do the following:

``` r
library(ggplot2)

ggplot(iris, aes(Sepal.Length, Sepal.Width)) +
  geom_point(aes(colour = Species)) +
  theme(legend.key.height = rel(0.5))
#> Error: Theme element `legend.key.height` must be an `unit` object.
```
<sup>Created on 2020-04-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

I think it would require changes at these locations.

* This check would have to allow `rel` classes for `unit` elements

https://github.com/tidyverse/ggplot2/blob/7e6d1259237d9a2b48b8156e62ada2c47c411f31/R/theme.r#L530-L533

* This might need an extra logical branch for `rel`-`rel` and `rel`-`unit` elements  

https://github.com/tidyverse/ggplot2/blob/7e6d1259237d9a2b48b8156e62ada2c47c411f31/R/theme.r#L665-L668

* Don't allow `rel` objects as 'root'-nodes in the element tree. I think the most appropriate spot to do this would be in `validate_element()`.

https://github.com/tidyverse/ggplot2/blob/40f67518a8d0ee8d7b6a1d1851c206656864fac8/R/theme-elements.r#L508

Thanks for considering!

### PR DESCRIPTION
**Pull request by teunbrand**
This PR aims to fix #3951.

Essentially, anywhere one can specify a `unit()` in the theme for an element that is *not* a root element, one can now use `rel()` to inherit from the parent `unit` object with modification.

In the example below:
* Legend keys are 2x as long as they're wide
* Top x-axis ticks are 2x as long as default
* Bottom x-axis ticks are 4x as long as default, because inheritance passes through 2 `<rel>` objects.

``` r
devtools::load_all("~/packages/ggplot2")
#> ℹ Loading ggplot2

ggplot(mpg, aes(displ, hwy, colour = factor(cyl))) +
  geom_point() +
  guides(x.sec = "axis") +
  theme(
    legend.key.height = rel(2),
    axis.ticks.length.x = rel(2),
    axis.ticks.length.x.bottom = rel(2)
  )
```

<sup>Created on 2023-08-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by teunbrand** [comment]
As a note: it also works on `margin()`s but I don't think any theme margin settings inherit something.

``` r
devtools::load_all("~/packages/ggplot2")
#> ℹ Loading ggplot2
combine_elements(rel(2), margin(5,5,5,5))
#> [1] 10points 10points 10points 10points
```

<sup>Created on 2023-09-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/theme.R [hunk 1]
```diff
@@ -724,6 +724,19 @@ combine_elements <- function(e1, e2) {
...
+    if (is.rel(e2)) {
+      return(rel(unclass(e1) * unclass(e2)))
+    }
+    # If e2 is a unit, return modified unit
+    if (is.unit(e2)) {
+      return(unclass(e1) * e2)
+    }
+    return(e1)
+  }
+
```
Comment (thomasp85) [inline R/theme.R:739]: We might want to build in a check for whether `e2` is numeric and apply rel to that, just for completeness

**Review by thomasp85 — APPROVED** [review]
Minor thing but otherwise good to go

File: R/theme.R (hunk 1 above)
Comment (teunbrand) [reply R/theme.R]: I was today years old when I learned:

``` r
library(grid)
is.numeric(unit(1, "npc"))
#> [1] TRUE
is.numeric(unit(1, "npc") + unit(1, "cm"))
#> [1] FALSE
```

<sup>Created on 2023-09-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Pull request merged**